### PR TITLE
Optimise sync

### PR DESCRIPTION
**Pull request by michaelfig**
This PR implements steps 1 and 2 of #1639.

### CODE REVIEW

**Comment by michaelfig** [comment]
Notably, the touch-after-sync is avoided by using the `m` flag to `tar`, which does the touch itself.

Maybe controversially, I added the `--no-same-owner` flag, which is the only thing that really makes sense to me when syncing a source file directly to a container.

File: pkg/skaffold/sync/kubectl/kubectl.go [hunk 1]
```diff
@@ -60,18 +58,30 @@ func (k *Syncer) Sync(ctx context.Context, s *sync.Item) error {
...
-	if _, ok := syncedDirs[dir]; !ok {
-		cmds = []*exec.Cmd{exec.CommandContext(ctx, "kubectl", "exec", pod.Name, "-c", container.Name, "-n", pod.Namespace, "--", "mkdir", "-p", dir)}
-		syncedDirs[dir] = struct{}{}
-	}
-	copy := exec.CommandContext(ctx, "kubectl", "cp", src, fmt.Sprintf("%s/%s:%s", pod.Namespace, pod.Name, dst), "-c", container.Name)
-	return append(cmds, copy)
+func copyFileFn(ctx context.Context, pod v1.Pod, container v1.Container, files map[string]string) []*exec.Cmd {
+	// Use "m" flag to touch the files as they are copied.
+	reader, writer := io.Pipe()
+	copy := exec.CommandContext(ctx, "kubectl", "exec", pod.Name, "--namespace", pod.Namespace, "-c", container.Name, "-i",
```
Comment (r2d4) [inline pkg/skaffold/sync/kubectl/kubectl.go]: Remove the tar after extraction?

Also TIL about `-m` or `--touch` flag. Cool

File: pkg/skaffold/sync/kubectl/kubectl.go [hunk 2]
```diff
@@ -60,18 +58,30 @@ func (k *Syncer) Sync(ctx context.Context, s *sync.Item) error {
...
+func deleteFileFn(ctx context.Context, pod v1.Pod, container v1.Container, files map[string]string) []*exec.Cmd {
+	// "kubectl" is below...
+	deleteCmd := []string{"exec", pod.Name, "--namespace", pod.Namespace, "-c", container.Name, "--", "rm", "-rf"}
+	args := make([]string, 0, len(deleteCmd)+len(files))
+	args = append(args, deleteCmd...)
+	for _, dst := range files {
+		args = append(args, dst)
+	}
+	delete := exec.CommandContext(ctx, "kubectl", args...)
 	return []*exec.Cmd{delete}
```
Comment (r2d4) [inline pkg/skaffold/sync/kubectl/kubectl.go]: Both copyFn and deleteFn can just return a single exec.Cmd? WDYT?

File: pkg/skaffold/sync/sync.go [hunk 3]
```diff
@@ -140,35 +140,32 @@ func Perform(ctx context.Context, image string, files map[string]string, cmdFn f
...
+					numSynced++
 				}
 			}
 		}
+	}
 
-		if len(synced) != len(files) {
-			return errors.New("couldn't sync all the files in " + ns)
-		}
+	if numSynced == 0 {
```
Comment (r2d4) [inline pkg/skaffold/sync/sync.go]: Not sure if this check makes sense anymore?

File: pkg/skaffold/sync/kubectl/kubectl.go (hunk 1 above)
Comment (michaelfig) [reply pkg/skaffold/sync/kubectl/kubectl.go]: The tar file is piped from Go's memory directly to `kubectl exec`'s stdin.  So there's nothing to remove.

File: pkg/skaffold/sync/kubectl/kubectl.go (hunk 2 above)
Comment (michaelfig) [reply pkg/skaffold/sync/kubectl/kubectl.go]: No harm in leaving them as slices, but I'm indifferent.

File: pkg/skaffold/sync/kubectl/kubectl.go (hunk 1 above)
Comment (r2d4) [reply pkg/skaffold/sync/kubectl/kubectl.go]: Ah missed that 👍

File: pkg/skaffold/sync/sync.go (hunk 3 above)
Comment (michaelfig) [reply pkg/skaffold/sync/sync.go]: I put this in to pass the "no copy" test at `sync_test.go`:410.  If you want, both this and the test can be removed.

**Pull request merged**